### PR TITLE
5 dreamerv2のactorモデルのバグ

### PR DESCRIPTION
**Pull request by pchan1111**
action_dist.log_prob(unscaled_actions)
の形が(B, 1)ではなく(B, )となりバグが発生していたので、
unsqueeze(-1)
を追加して(B, 1)に形を変更してバグを修正しました！

### CODE REVIEW

File: Dreamer_V2.ipynb [hunk 1]
```diff
@@ -281,7 +288,7 @@
         "                次時刻の決定的状態h_t+1\n",
         "        \"\"\"\n",
         "        mlp_hidden = F.elu(self.z_a_hidden(torch.cat([z_prev, a_prev], dim=1)))\n",
-        "        h = self.rnn(mlp_hidden, rnn_hidden)\n",
+        "        h = self.rnn(mlp_hidden, mlp_hidden)\n",
```
Comment (seki-shu) [inline Dreamer_V2.ipynb]: `h = self.rnn(mlp_hidden, h_prev)` ではないですか？引数のh_prev が使われてないです．

File: Dreamer_V2.ipynb [hunk 2]
```diff
@@ -978,6 +986,565 @@
...
+        "\n",
+        "        # priorを用いた状態予測\n",
+        "        # 格納する空のtensorを用意\n",
+        "        imagined_z = torch.zeros(cfg.imagination_horizon + 1,\n",
+        "                                 *z_flat.shape,\n",
+        "                                 device=z_flat.device) # (imagination_horizon + 1, (T-1) * B, z_dim * n_classes)\n",
+        "        imagined_h = torch.zeros(cfg.imagination_horizon + 1,\n",
+        "                                 *h_flat.shape,\n",
+        "                                 device=h_flat.device) # (imagination_horizon + 1, (T-1) * B, h_dim)\n",
+        "        imagined_action_log_probs = torch.zeros(cfg.imagination_horizon,\n",
```
Comment (seki-shu) [inline Dreamer_V2.ipynb:1429]: ここで期待している形`(imagination_horizon, (T-1) * B)`から以下で期待される形は((T-1) * B)になりますが，action_log_probsの形が異なります．
`imagined_action_log_probs[i] = action_log_probs`

File: Dreamer_V2.ipynb [hunk 3]
```diff
@@ -978,6 +986,565 @@
...
+        "        # 想像上の軌道を作る前に、最初の状態として先ほどReplayBufferからサンプルされた観測データを取り込む\n",
+        "        imagined_z[0] = z_flat # ((T-1) * B, z_dim * n_classes)\n",
+        "        imagined_h[0] = h_flat # ((T-1) * B, h_dim)\n",
+        "\n",
+        "        # open-loopで予測\n",
+        "        # まず世界モデル上で想像上の状態、行動を取得\n",
+        "        # その後それらにすべてに対してまとめて報酬、割引率（終了判定)、状態価値を予測\n",
+        "        # 想像上の状態、行動生成\n",
+        "        for i in range(cfg.imagination_horizon):\n",
+        "            actions, action_log_probs, action_entropies = actor(imagined_h[i], imagined_z[i]) # ((T-1) * B, action_dim or 1)\n",
```
Comment (seki-shu) [inline Dreamer_V2.ipynb:1445]: `action_log_probs` を((T-1) * B,1)の形になっています．((T-1) * B)にしたいです．

File: Dreamer_V2.ipynb [hunk 4]
```diff
@@ -978,6 +986,565 @@
...
+        "        # すべての観測をベクトルに埋め込み\n",
+        "        emb_observations = encoder(observations.reshape(-1, 1, 64, 64)).view(cfg.seq_length, cfg.batch_size, -1) # (T, B, 1536)\n",
+        "        # 各時刻tごとのz,hを保存するためのtensorを定義\n",
+        "        z = torch.zeros(cfg.seq_length, cfg.batch_size, cfg.z_dim * cfg.n_classes, device=device)\n",
+        "        h = torch.zeros(cfg.seq_length, cfg.batch_size, cfg.h_dim, device=device)\n",
+        "        # KL lossの計算\n",
+        "        kl_loss = 0\n",
+        "        for t in range(cfg.seq_length-1):\n",
+        "            # hの更新\n",
+        "            h[t+1] = rsssm.recurrent(h[t], z[t], actions[t]) # h_t+1\n",
```
Comment (seki-shu) [inline Dreamer_V2.ipynb:1379]: ここですぐにh[t+1]を更新してしまうと，次の`prior, detached_prior = rssm.get_prior(h[t+1], detach=True)`で更新された値が使われてしまうので，バグになります．

**Review by seki-shu — COMMENTED** [review]
いくつかのバグをコメントしました．意図と合っていないところがあれば教えてください．

File: Dreamer_V2.ipynb (hunk 1 above)
Comment (pchan1111) [reply Dreamer_V2.ipynb]: h_prevに修正しました！

File: Dreamer_V2.ipynb (hunk 2 above)
Comment (pchan1111) [reply Dreamer_V2.ipynb:1429]: 修正しました！それに伴って、actorのところではaction_log_probsの形状が(B, 1)を返すとしていましたが、(B, )を返すの間違いでしたので変更しました

File: Dreamer_V2.ipynb (hunk 3 above)
Comment (pchan1111) [reply Dreamer_V2.ipynb:1445]: 上の修正の通りです！

File: Dreamer_V2.ipynb (hunk 4 above)
Comment (pchan1111) [reply Dreamer_V2.ipynb:1379]: ここの部分は配布された資料とほぼ同じで、h(t), z(t), a(t)が与えられたらh(t+1)を計算してそれを基にpriorを計算するので多分大丈夫なはずです！追加で質問があれば教えて下さい。
